javascript/ql/src/Security/CWE-601/examples/ServerSideUrlRedirectGood2.js

```diff
@@ -0,0 +1,15 @@
+const app = require("express")();
+
+function isRelativePath(path) {
+  return !/^(\w+:)?[/\\]{2}/.test(path);
```

Comment:
```suggestion
  return !/^([\w+-.]+:)?[/\\]{2}/.test(path);
```
`+`, `-`, and `.` are also allowed in the scheme, so the final regexp ends up looking a little ugly.

From https://www.ietf.org/rfc/rfc1738.txt:

> Scheme names consist of a sequence of characters. The lower case
   letters "a"--"z", digits, and the characters plus ("+"), period
   ("."), and hyphen ("-") are allowed.

Alternatively, one could parse the URL with a hypothetical base origin and check that the origin was unchanged (which works regardless of what the actual hostname of the server is):
```js
  try {
    return new URL(path, 'https://example.com').origin === 'https://example.com';
  } catch (e) {
    return false;
  }
```

Reply:
This is a beautiful idea! So much neater than regex faffery. Updated accordingly.

Reply:
Of course, it does allow redirects to example.com, which is harmless but perhaps unexpected. Perhaps best to mention this somewhere?